### PR TITLE
add second fraction normalisation

### LINKED ISSUE #702: `TimeLike` does not correct account for nanoseconds exceeding 1e9
I'm been using chrono as a reference implementation while fuzzing my own code, and I've found an error in chrono.

`.to_string()` correctly rounds seconds up to account for nanoseconds exceeding 1e9, but `.second()` and `.nanosecond()` from `TimeLike` do not.

Example:

```rs
use chrono::{NaiveDate, Timelike};

#[test]
fn chrono_error() {
    let chrono_dt = NaiveDate::from_ymd(2000, 1, 1).and_hms_nano(1, 0, 0, 1_010_000_000);
    assert_eq!(chrono_dt.to_string(), "2000-01-01 01:00:01.010");
    assert_eq!((chrono_dt.second(), chrono_dt.nanosecond()), (1, 10_000_000));
}
```

gives:

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `(0, 1010000000)`,
 right: `(1, 10000000)`', src/main.rs:6:5
```

[rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fa333040c21e4cc1ab406911254390fb) showing the error.

### PR DESCRIPTION
**Pull request by samuelcolvin**
fix #702

- [x] Have you added yourself and the change to the [changelog]? (Don't worry
      about adding the PR number)
- [x] If this pull request fixes a bug, does it add a test that verifies that
      we can't reintroduce it?

[changelog]: ../CHANGELOG.md

### CODE REVIEW

File: src/naive/time/tests.rs [hunk 1]
```diff
@@ -276,3 +276,17 @@ fn test_time_format() {
...
+
+
+#[test]
+fn test_timelike_normalisation() {
+    let t = NaiveTime::from_hms_nano(0, 0, 0, 1_123_456_789);
+    assert_eq!(t.second(), 1);
+    assert_eq!(t.nanosecond(), 123_456_789);
+
+    let t = NaiveTime::from_hms_nano(23, 59, 59, 1_999_999_999);
+    assert_eq!(t.hour(), 24);
```
Comment (samuelcolvin) [inline src/naive/time/tests.rs:288]: I don't know if this is correct, the alternative is to alter `from_hms_nano_opt` and `with_nanosecond` to error if the total time exceeds 24hours which seems like a bigger change.

If we want to keep the change this way, I think better to be explicit about this with a test.

**Comment by samuelcolvin** [comment]
Humm, I'm not sure what to do about these tests - it seems like to support leap seconds, you explicitly want the functionality that this PR alters?

**Comment by djc** [comment]
Ah yeah, I don't have enough context on that -- I only became a maintainer of this crate recently.

File: src/naive/time/tests.rs (hunk 1 above)
Comment (esheppa) [reply src/naive/time/tests.rs:287]: In this case `.second()` should return 60 due to the nanosecond representation being equal to or greater than 1e9 which represents a leap second in `chrono`, but `.hour()` should still return 23

**Comment by esheppa** [comment]
Hi @samuelcolvin - I've made some comments over at https://github.com/chronotope/chrono/issues/702 - please let me know your thoughts

**Comment by pitdicker** [comment]
Closing due to inactivity and the concerns raised above. We still have https://github.com/chronotope/chrono/issues/702 to track this issue.

**Pull request closed without merging**